java/org/apache/catalina/realm/GenericPrincipal.java

```diff
@@ -171,6 +176,7 @@ public GenericPrincipal(String name, List<String> roles,
         }
         this.loginContext = loginContext;
         this.gssCredential = gssCredential;
+        this.attributes = attributes;
```

Comment:
Should we wrap with an unmodifiable map or do we expect the caller to do this?

Reply:
AFAIK we agreed on not using an unmodifiable map, since there is no write access to this map. However, we could, of course, if you like so. Maybe it's worth checking if it's already an instance of `java.util.Collections.UnmodifiableMap` first. If not, we could wrap it.

Reply:
Class `java.util.Collections.UnmodifiableMap` is private, so we cannot check instance.